core/src/main/java/cucumber/runtime/Runtime.java

```diff
@@ -73,15 +75,17 @@ public void run() {
 
         final FeatureCompiler compiler = new FeatureCompiler();
         for (CucumberFeature feature : features) {
-            for (final PickleEvent pickleEvent : compiler.compileFeature(feature)) {
-                if (filters.matchesFilters(pickleEvent)) {
+            final List<PickleEvent> matchedEvents = getMatchingPickles(compiler.compileFeature(feature));
+
+            bus.send(new FeatureTestCaseExpectationEvent(bus.getTime(), feature.getUri(), matchedEvents.size()));
```

Comment:
This won't work in JUnit and TestNG. Cucumber filters pickles, then JUnit filters test cases created from those pickles.

Reply:
JUnit and TestNG call through `cucumber.runner.Runner#runPickle` not `cucumber.runtime.Runtime#run`

Only `Main` calls into `cucumber.runtime.Runtime#run`

Reply:
> they also do not use OrderedTestRunEventBus

Mmh. Formatters can also listen to JUnit and TestNG so they should be using OrderedTestRunEventBus. Must have missed it when adding the bus.

Reply:
Listen or register?  Doesn't that contradict your earlier comment/feedback

```
Junit Thread 1 ---listens to---> [TestCase EventBus for Runner 1] ---sends batches of events to--> [Synchronized EventBus]  <---listens too--- HTML formatter 
 Junit Thread 2 ---listens to---> [TestCase EventBus for Runner 2] ---sends batches of events to--> |
```

Reply:
Sorry. Being sloppy with my language. Formatters also subscribe to the event bus during a JUnit/TestNG execution of cucumber. These may also execute pickles in any order. 

Currently the `OrderedTestRunEventBus` comes after the synchronized bus. Like so:

```
Junit Thread 1 ---listens to---> [TestCase EventBus for Runner 1] ---sends batches of events to--> [Synchronized EventBus]  --Sends events to--> [OrderedTestRunEventBus]   <---listens too batches --- HTML formatter 
Junit Thread 2 ---listens to---> [TestCase EventBus for Runner 2] ---sends batches of events to--> |
```

Where I want to go is:

```
Junit Thread 1 ---listens to---> [TestCase EventBus for Runner 1] ---sends events to--> [Synchronized EventBus]  <---listens too--- HTML formatter  
Junit Thread 2 ---listens to---> [TestCase EventBus for Runner 2] ---sends events to--> |
```

The TestCase EventBus no longer sends batches but passes events to it's listeners and parent directly. The HTML formatter gets them as we go along. It is then the HTML formatters responsibility to handle events from parallel  execution correctly. We can however make this easier by providing a comparator or implement comparable as part of the public api.